src/main/java/org/opentripplanner/middleware/controllers/api/ApiUserController.java

```diff
@@ -1,28 +1,170 @@
 import org.opentripplanner.middleware.auth.Auth0UserProfile;
-import org.opentripplanner.middleware.utils.ApiGatewayUtils;
+import org.opentripplanner.middleware.models.ApiKey;
 import org.opentripplanner.middleware.models.ApiUser;
 import org.opentripplanner.middleware.persistence.Persistence;
+import org.opentripplanner.middleware.utils.ApiGatewayUtils;
+import org.opentripplanner.middleware.utils.HttpUtils;
+import org.opentripplanner.middleware.utils.JsonUtils;
 import spark.HaltException;
 import spark.Request;
+import spark.Response;
+
+import java.util.Collections;
 
+import static com.beerboy.ss.descriptor.MethodDescriptor.path;
 import static org.opentripplanner.middleware.OtpMiddlewareMain.getConfigPropertyAsText;
+import static org.opentripplanner.middleware.auth.Auth0Connection.isUserAdmin;
 import static org.opentripplanner.middleware.utils.JsonUtils.logMessageAndHalt;
 
 /**
- * Implementation of the {@link AbstractUserController} for {@link ApiUser}.
+ * Implementation of the {@link AbstractUserController} for {@link ApiUser}. This controller also contains methods for
+ * managing an {@link ApiUser}'s API keys.
  */
 public class ApiUserController extends AbstractUserController<ApiUser> {
-    private static final String DEFAULT_USAGE_PLAN_ID
-        = getConfigPropertyAsText("DEFAULT_USAGE_PLAN_ID");
+    private static final String DEFAULT_USAGE_PLAN_ID = getConfigPropertyAsText("DEFAULT_USAGE_PLAN_ID");
+    private static final String API_KEY_PATH = "/apikey";
+    private static final int API_KEY_LIMIT_PER_USER = 2;
+    private static final String API_KEY_ID_PARAM = "/:apiKeyId";
 
     public ApiUserController(String apiPrefix) {
         super(apiPrefix, Persistence.apiUsers, "secure/application");
     }
 
+    @Override
+    protected void buildEndpoint(ApiEndpoint baseEndpoint) {
+        LOG.info("Registering path {}.", ROOT_ROUTE + API_KEY_PATH);
+
+        // Add the api key route BEFORE the regular CRUD methods
+        // TODO: Define usage plan id as an optional parameter under Spark Swagger. 'withParam' method does not register
+        //  the parameter and 'withPathParam' is not optional (even with 'withRequired(false))'. For the time being
+        //  appending '?usagePlanId=<id>' works.
+        ApiEndpoint modifiedEndpoint = baseEndpoint
+            // Reveal the actual API key value for the key ID
+            .get(path(ROOT_ROUTE + API_KEY_PATH + API_KEY_ID_PARAM)
+                    .withDescription("Gets the API key value for a given api key ID")
+                    .withPathParam().withName("apiKeyId").withDescription("The api key ID of the API key.").and()
+                    .withResponseType(ApiKey.class),
+                this::getApiKey, JsonUtils::toJson
+            )
+            // Create API key
+            .post(path(API_KEY_PATH)
+                    .withDescription("Creates API key for ApiUser (with optional usagePlanId)")
+                    .withResponseType(persistence.clazz),
```

Comment:
Also insert the query param for `userId`.

Reply:
I am leaning towards a URL scheme like `api/secure/application/:userid/apikey` would be more suitable for this POST method and `api/secure/application/:id/apikey/:apikeyid` for the DELETE method. I think we should adopt that unless there is a strong case against it.

Reply:
I've inserted the query params for usagePlanId and userId.

@binh-dam-ibigroup The ApiUser would have to define their userId instead of it being obtained from their token. The userId in the URL would still need to be checked against this. I think the URL scheme approach is more suited to an admin user where the userId is required (at least for POST). I'm not against making the change, but would like agreement from @landonreed and @evansiroky before updating and testing it.

Reply:
I think this URL scheme makes sense. I'm down with this approach.

Reply:
@br648, I think we should be importing the param/path field from ApiController, similar to what we're doing in #33 https://github.com/ibi-group/otp-middleware/pull/33/files#diff-3772d28a6571870eef9631b391ccc391R41-R42

Reply:
Updated.